contracts/finality/src/state/rate_limiting.rs

```diff
@@ -0,0 +1,315 @@
+use crate::{error::ContractError, state::config::get_config};
+use cosmwasm_std::{Env, Storage};
+
+use cw_storage_plus::Map;
+
+/// Stores the number of messages processed for each finality provider within the last block interval
+/// Key: Finality Provider's BTC public key
+/// Value: Tuple of (block interval number, message count)
```

Comment:
Why do we need the block interval number?

Reply:
Your answer here seems to be related to the message count though. What I'm mostly curious about is why we need to keep a track of messages for past block intervals instead of just the current one.

Reply:
> why we need to keep a track of messages for past block intervals instead of just the current one.

I think the current implementation only tracks the current one. So the DB schema is

```rust
const NUM_MSGS_LAST_INTERVAL: Map<&[u8], (u64, u32)> = Map::new("num_msgs_last_interval");
```

where the key is just the FP PK and value contains the interval number and count. Any write operation to this KV store for a given FP will overwrite the previous record for this FP.